src/pubsub/src/publisher/worker.rs

```diff
@@ -92,40 +98,48 @@ impl Worker {
-                    batch.flush(self.client.clone(), self.topic_name.clone(), &mut inflight);
+                _ = &mut timer => {
+                    for (_, outstanding) in pending_batches.iter_mut() {
+                        if !outstanding.pending_batch.is_empty() {
+                            outstanding.pending_batch.flush(self.client.clone(), self.topic_name.clone(), &mut inflight);
+                        }
+                    }
+                    timer.as_mut().reset(tokio::time::Instant::now() + delay);
                 }
                 // Handle receiving a message from the channel.
                 msg = self.rx.recv() => {
                     match msg {
                         Some(ToWorker::Publish(msg)) => {
-                            // Reset the timer if this is the first message to be added to the batch.
-                            if batch.is_empty() {
-                                timer.as_mut().reset(tokio::time::Instant::now() + delay);
-                            }
-                            batch.push(msg);
-                            if batch.len() as u32 >= message_limit || batch.size() >=  byte_threshold {
-                                batch.flush(self.client.clone(), self.topic_name.clone(), &mut inflight);
+                            let ordering_key = msg.msg.ordering_key.clone();
+                            let outstanding_publishes = pending_batches.entry(ordering_key).or_insert(OutstandingPublishes::new(&self.topic_name));
+                            outstanding_publishes.pending_batch.push(msg);
+                            if outstanding_publishes.pending_batch.len() as u32 >= message_limit || outstanding_publishes.pending_batch.size() >=  byte_threshold {
+                                outstanding_publishes.pending_batch.flush(self.client.clone(), self.topic_name.clone(), &mut inflight);
                             }
                         },
                         Some(ToWorker::Flush(tx)) => {
-                            batch.flush(self.client.clone(), self.topic_name.clone(), &mut inflight);
-                            // Wait on all the tasks that exist right now.
-                            // We could instead tokio::spawn this as well so the publisher
-                            // can keep working on additional messages. The worker would
-                            // also need to keep track of any pending flushes, and make sure
-                            // all of those resolve as well.
-                            let mut flushing = std::mem::take(&mut inflight);
-                            while flushing.next().await.is_some() {}
+                            for (_, outstanding) in pending_batches.iter_mut() {
+                                if !outstanding.pending_batch.is_empty() {
+                                    outstanding.pending_batch.flush(self.client.clone(), self.topic_name.clone(), &mut inflight);
```

Comment:
I would at least add a comment / TODO here for the behavior of flush when there are ordering keys. This will involve sending multiple batches *sequentially* since they need to be ordered. So it is not as simple as just flushing the existing messages and waiting on the response.